### PR TITLE
All migrations tests now async/await'd and passing

### PR DESCRIPTION
**Pull request by zemccartney**
Makes adjustments necessary to get all migrations tests passing

### CODE REVIEW

**Comment by devinivy** [comment]
A total bombshell!  Sick– I look forward to digging thru this.  Anything in particular you're feeling funny about and want me to checkout?

File: lib/migrator.js [hunk 1]
```diff
@@ -48,27 +68,32 @@ internals.executeMigration = async (migration, rollback) => {
...
 };
 
 internals.readMigrationDir = async (migrationDir, symlinkConfig) => {
 
-    const migrationFiles = await internals.PROMISIFIED_UTILITIES.readDirAsync(migrationDir);
+    const migrationFiles = await Util.promisify(Fs.readdir)(migrationDir);
+
+    // We store our current migrationDir here and not in the above for...of loop in executeMigration because if assigned there, subsequent iterations of that loop would overwrite the copy of symlinkConfig that the current iteration of readMigrationDir is reading and using to create symlinks
+    // So, the created symlinks would use the files read from the current migrationDir, but then the migrationDir from the next iteration, resulting in ENOENT errors
+    symlinkConfig.migrationDir = migrationDir;
```
Comment (devinivy) [inline lib/migrator.js]: Typically if you're mutating input arguments there's a better way out there!

File: lib/migrator.js [hunk 2]
```diff
@@ -48,27 +68,32 @@ internals.executeMigration = async (migration, rollback) => {
...
+
     const migrationSymlinks = [];
 
-    for (const file in migrationFiles) {
+    for (const file of migrationFiles) {
```
Comment (devinivy) [inline lib/migrator.js]: Is this fixing a previous change you made, or is this from the original migration code?

File: lib/migrator.js [hunk 3]
```diff
@@ -81,43 +106,36 @@ internals.symlinkMigrationFile = async (migrationFile, symlinkConfig) => {
...
+ *
+ * @param {object} opts — Options object that Tmp.dir normally expects
+ * @return {promise} Promise that, if Tmp.dir succeeds, fulfills with an object containing
+ *      - *tmpPath* the path to the new temp directory
+ *      - *cleanupCallback* {function} removes the temporary directory when we're done using it
+ */
+internals.promisifyTmpDir = (opts) => {
 
-        Hoek.assert(knex, 'Cannot specify a migrations directory without an available knex instance.');
+    return new Promise((fulfill, reject) => {
```
Comment (devinivy) [inline lib/migrator.js]: Old school!  `fulfill` is better known today as `resolve`, though they are pretty much synonymous.  To do the "standard" thing, can we change this argument to `resolve`, though?

File: lib/migrator.js [hunk 4]
```diff
@@ -81,43 +106,36 @@ internals.symlinkMigrationFile = async (migrationFile, symlinkConfig) => {
...
 
-        if (!knexMigrations.has(knex)) {
-            knexMigrations.set(knex, {});
-        }
+        Tmp.dir(opts, (err, path, cleanupCallback) => {
 
-        const migrationsDirs = knexMigrations.get(knex);
+            if (err) {
+                throw err;
```
Comment (devinivy) [inline lib/migrator.js]: If you want to be explicit about it, `return reject(err)`.  I _think_ this works too, though?

File: lib/migrator.js (hunk 1 above)
Comment (zemccartney) [reply lib/migrator.js]: ay! shit, you just taught me about this. dag! thanks for pointing out, I'll dig at it again 🙏

File: lib/migrator.js (hunk 2 above)
Comment (zemccartney) [reply lib/migrator.js]: fixing a previous change. I think writing `in` was just a (subtly unfortunate 🎺 ) typo on my part that caused things to break.

File: lib/migrator.js (hunk 3 above)
Comment (zemccartney) [reply lib/migrator.js]: Ahhhhh, gotcha. Wicked, thanks for pointing that out. For sure, I'll fix that 🙏

File: lib/migrator.js (hunk 4 above)
Comment (zemccartney) [reply lib/migrator.js]: Oh man, I can never keep straight how error throwing works in Promises / I think it's like different in then calls vs. the originating Promise? Woof..... So yea! Explicit definitely sounds like the move. Thanks for that, homie

**Comment by zemccartney** [comment]
💋  Thanks for the review, my dude, appreciate all the help (now and so far) 🙏 

Hairy spots that come to mind

- manually promisifying the Tmp.dir function; I had a whole slew of trouble with and making sure I exposed, then used, the cleanup callback properly; though I think you've covered that already 🙏 
- executing groups of Promises in sequence vs. in parallel in the migrator. I think I did this right and ordered things correctly? But man, it took a chunk of a day and my brain to sort out how to order and compose the various methods used in exports.migrate
- the `'coalesces migrations in different directories across plugins sharing knex instances.'` ... I think it's ok, just took me a while to understand that one enough to rewrite, so seems a likely spot for hiccups.

**Comment by zemccartney** [comment]
@devinivy yo! I _think_ I got everything, but lemme know. No urgency here, though, I'm all good to keep plugging away at the rest of the upgrade 🙏  thanks for all the notes so far

File: lib/migrator.js [hunk 5]
```diff
@@ -81,43 +106,36 @@ internals.symlinkMigrationFile = async (migrationFile, symlinkConfig) => {
...
+    return new Promise((resolve, reject) => {
 
-        if (!knexMigrations.has(knex)) {
-            knexMigrations.set(knex, {});
-        }
+        Tmp.dir(opts, (err, path, cleanupCallback) => {
 
-        const migrationsDirs = knexMigrations.get(knex);
+            if (err) {
+                reject(err);
```
Comment (devinivy) [inline lib/migrator.js]: Needs a `return` here.

File: lib/migrator.js [hunk 6]
```diff
@@ -48,27 +68,32 @@ internals.executeMigration = async (migration, rollback) => {
...
     await knex.migrate[latestOrRollback]({ directory: relMigrationsDir });
 
-    // TODO does this work? What happens when a callback takes 2 non-error parameters?
-    tmpResults.cleanup();
+    tmpCleanup();
+
 };
 
-internals.readMigrationDir = async (migrationDir, symlinkConfig) => {
+internals.readMigrationDir = async (migrationDir, symlinkStore) => {
```
Comment (devinivy) [inline lib/migrator.js]: This function does read directories, but the important action it takes is really symlinking each file in a migration directory to a file in a temp directory.

File: lib/migrator.js [hunk 7]
```diff
@@ -4,42 +4,62 @@ const Fs = require('fs');
...
-    const tmpResults = internals.PROMISIFIED_UTILITIES.tempDirAsync({ unsafeCleanup: true });
-    console.log(tmpResults, 'BOLOGNESE');
-    // iT'S PROBABLY AN ARRAY (OR ARRAY-LIKE OBJECT E.G. ARGUMENTS)
+    // Promisifying callback-style functions at runtime allows us to monkey-patch these methods in our tests to simulate errors, see /test/index.js, migrations tests usage of Fs.readdir and Tmp.dir
+    const { tmpPath, tmpCleanup } = await internals.promisifyTmpDir({ unsafeCleanup: true });
 
     const directoriesToRead = [];
-    const symlinkConfig = {
-        tmpPath: tmpResults[0]
+    const symlinkStore = {
```
Comment (devinivy) [inline lib/migrator.js]: Could the `tmpPath` just be passed around rather than wrapped in an object?

File: lib/migrator.js [hunk 8]
```diff
@@ -48,27 +68,32 @@ internals.executeMigration = async (migration, rollback) => {
...
+        tmpCleanup();
+        throw err;
     }
 
-    // TODO Review / fix tmpAssets.tmpPath
-    const relMigrationsDir = Path.relative(process.cwd(), tmpResults.tmpPath);
+    const relMigrationsDir = Path.relative(process.cwd(), tmpPath);
     const latestOrRollback = rollback ? 'rollback' : 'latest';
 
     await knex.migrate[latestOrRollback]({ directory: relMigrationsDir });
```
Comment (devinivy) [inline lib/migrator.js]: If this line throws, will `tmpCleanup()` be called?  You might want a `finally {}` block for pretty much this whole function.

File: lib/migrator.js [hunk 9]
```diff
@@ -81,43 +106,36 @@ internals.symlinkMigrationFile = async (migrationFile, symlinkConfig) => {
...
+ * A non-standard one :) Wraps the Tmp.dir method in a promise. We do this manually instead of using Util.promisify as we do with the other callback-style methods
+ * because Tmp.dir's callback takes 2 non-error parameters, but Promises fulfill with only a single value and Util.promisify handles this by
+ * fulfilling with only the first parameter (the path to the temporary directory), which resulted in us being unable to cleanup our temporary directory because we no longer had access to Tmp's cleanup method
+ *
+ * @param {object} opts — Options object that Tmp.dir normally expects
+ * @return {promise} Promise that, if Tmp.dir succeeds, fulfills with an object containing
+ *      - *tmpPath* the path to the new temp directory
+ *      - *cleanupCallback* {function} removes the temporary directory when we're done using it
+ */
+internals.promisifyTmpDir = (opts) => {
```
Comment (devinivy) [inline lib/migrator.js]: If you like this name no problem, but it could also just be `internals.tmpDir`.

File: lib/migrator.js [hunk 10]
```diff
@@ -81,43 +106,36 @@ internals.symlinkMigrationFile = async (migrationFile, symlinkConfig) => {
...
-        }
+};
 
-        const knex = knexGroup.knex || defaultKnex;
+/**
+ * A non-standard one :) Wraps the Tmp.dir method in a promise. We do this manually instead of using Util.promisify as we do with the other callback-style methods
+ * because Tmp.dir's callback takes 2 non-error parameters, but Promises fulfill with only a single value and Util.promisify handles this by
+ * fulfilling with only the first parameter (the path to the temporary directory), which resulted in us being unable to cleanup our temporary directory because we no longer had access to Tmp's cleanup method
+ *
+ * @param {object} opts — Options object that Tmp.dir normally expects
```
Comment (devinivy) [inline lib/migrator.js]: The comment is fine, but can we get rid of the doc block?  I find it's all or nothing with those suckers.

File: package.json [hunk 11]
```diff
@@ -28,12 +28,12 @@
     "toys": "1.x.x"
   },
   "peerDependencies": {
-    "hapi": ">=10 <=17",
+    "hapi": ">=17",
```
Comment (devinivy) [inline package.json]: And also `<18`!  Can be written `17.x.x`.

**Comment by devinivy** [comment]
@zemccartney did another run through.  There are definitely a few things left, but overall this is looking pretty dang good!  Crusht the tests.

**Comment by zemccartney** [comment]
@devinivy thanks for all the notes, my dude! just pushed things up, I think handled everything 🙏

File: package.json [hunk 12]
```diff
@@ -28,7 +28,7 @@
     "toys": "1.x.x"
   },
   "peerDependencies": {
-    "hapi": ">=17",
+    "hapi": ">=17.x.x <18",
```
Comment (devinivy) [inline package.json]: Or just `"hapi": "17.x.x"`

**Review by devinivy — APPROVED** [review]
:eagle: 
🙌

**Comment by zemccartney** [comment]
🙏 💋

**Comment by wswoodruff** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Pull request merged**